ext/drreg/drreg.c

```diff
@@ -2107,80 +2107,26 @@ drreg_event_restore_state_with_ilist(void *drcontext, bool restore_memory,
-        if (!instr_is_app(inst) &&
-            is_our_spill_or_restore(drcontext, inst, &spill, &reg, &slot, &offs)) {
-            if (spill) {
-                /* Slot used in current instr may have a previous spilled gpr/aflags
-                 * value that may already have been restored. Forget that.
-                 */
-                spill_slot_to_reg[slot] = DR_REG_NULL;
-                if (aflags_spill_reg == reg) {
-                    ASSERT(!gpr_native[GPR_IDX(aflags_spill_reg)],
-                           "reg with aflags cannot be native");
-                    spill_slot_to_reg[slot] = AFLAGS_ALIAS_REG;
-                } else if (gpr_native[GPR_IDX(reg)]) {
-                    spill_slot_to_reg[slot] = reg;
-                } else {
-                    LOG(drcontext, DR_LOG_ALL, 3,
-                        "%s @" PFX ": ignoring tool spill of non-native value.\n",
-                        __FUNCTION__, pc);
-                }
-            } else {
-                if (spill_slot_to_reg[slot] == AFLAGS_ALIAS_REG) {
-                    aflags_spill_reg = reg;
-                    app_aflags_written_to_reg_now = true;
-                } else if (spill_slot_to_reg[slot] == reg) {
-                    gpr_native[GPR_IDX(reg)] = true;
-                    app_gpr_restored_now = true;
-                } else {
-                    LOG(drcontext, DR_LOG_ALL, 3,
-                        "%s @" PFX ": ignoring tool restore of non-native value\n",
-                        __FUNCTION__, pc);
-                }
-            }
-        } else if (!instr_is_app(inst) &&
-                   instr_get_opcode(inst) == IF_X86_ELSE(OP_lahf, OP_mrs)) {
-            if (aflags_native) {
-#ifdef X86
-                aflags_spill_reg = DR_REG_XAX;
-#else
-                aflags_spill_reg = opnd_get_reg(instr_get_dst(inst, 0));
-#endif
-                app_aflags_written_to_reg_now = true;
-            }
-        } else if (!instr_is_app(inst) &&
-                   instr_get_opcode(inst) == IF_X86_ELSE(OP_sahf, OP_msr)) {
-            if (aflags_spill_reg ==
-                IF_X86_ELSE(DR_REG_XAX, opnd_get_reg(instr_get_src(inst, 0)))) {
-                aflags_native = true;
-                app_aflags_restored_now = true;
-            }
-        }
-
         /* Mark gprs/aflags as containing native or non-native value based on whether
          * the current instr is an app or tool write, respectively. An example of the
          * latter is a restore for a spilled tool value, which may happen in the
          * multi-phase nested reservation case.
-         * If the current instr is a tool write that restores an app value (as tracked
-         * by the various *_now bools), we skip marking the gpr/aflags as non-native.
+         * Later we check whether the current instr restores native app value for any
+         * gpr/aflags, and if it does we mark the written gpr/aflags as native.
```

Comment:
This sounds like a behavior change, and not just refactoring?

Before: "we skip marking as non-native" for "a tool write that restores".
Now: we *do* mark as native.

Yet the PR description just says it's a (non-semantic-changing) refactor?

Reply:
The final outcome is the same. Although we do mark a restored gpr as non-native initially, but the "is this a restore?" check (which is moved to later in this PR) will eventually mark it as native; this is described in a comment there. This is simpler as it avoids any need of the bool flag (the *_now variables). There is no correctness issue due to this reordering because for restore instrs, `gpr_native` is not read.

Before:
<mark restored gpr as native; set flag F>
<unless F set, if meta instr, mark all dest gpr as non-native>

Now:
<if meta instr, mark all dest gpr as non-native>
<mark restored gpr as native>

Reply:
Similar reasoning shows that the final outcome for other vars (`aflags_native` and `aflags_reg`) is also the same.

Reply:
OK.